Add draco mesh compression

FrankGalligan opened the pull request:
The Draco glTF files created with gltf-pipeline will be created according to the KHR_draco_mesh_compression specification located
here:
https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_draco_mesh_compression/README.md

lilleyse commented:
Awesome @FrankGalligan and @fanzhanggoogle. One of us should have a chance to review pretty soon.

Is it possible to clean up the git history a bit?

lilleyse commented:
Fixes #348

shehzan10 commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+              console.log('Found duplicate!');
+              // Copy compressed primitive.
+              copyCompressedExtensionToPrimitive(
+                  primitive, hashPrimitives[hashValue]);
```
Formatting: Combine into same line as the function.

FrankGalligan replied:
Done.

shehzan10 commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+            const numFaces = indices.length / 3;
+            const numIndices = indices.length;
+
+            console.log("Num of faces: " + numFaces);
```
This should probably be removed.

FrankGalligan replied:
Done.

lilleyse commented on lib/getComponentReader.js:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @private
+ */
+function getComponentReader(componentType) {
```
Actually the `2.0` branch doesn't have any unit tests now... but they will come before it is merged into master.

FrankGalligan commented:
@lilleyse For the history let me see what I can do. I tried to pull your 2.0 branch into my fork and then my "add_draco_mesh_compression" branch on top of that. But this PR does not reflect the history I would have liked.

lilleyse commented on lib/draco_encoder_nodejs.js:
```diff
@@ -0,0 +1,28 @@
+module.exports = CreateDracoModule;
```
Unless I'm missing something it looks like the JS encoder isn't being used.

FrankGalligan replied:
True. Removed this file.

shehzan10 commented on README.md:
```diff
@@ -105,6 +108,13 @@ processGltf(gltf, options)
 |`--separateTextures`, `-t`|Write out separate textures only.|No, default `false`|
 |`--checkTransparency`|Do a more exhaustive check for texture transparency by looking at the alpha channel of each pixel. By default textures are considered to be opaque.|No, default `false`|
 |`--stats`|Print statistics to console for input and output glTF files.|No, default `false`|
+|`--draco.compressionLevel`|Draco compression level [0-10], most is 10, least is 0.|No, default `7`|
```
Need to add `compressMesh` option.here.

FrankGalligan replied:
Done.

FrankGalligan replied:
Done.

shehzan10 commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+            if (encodedLen > 0) {
+              console.log("Encoded size is " + encodedLen);
+            } else {
+              console.log("Error: Encoding failed.");
```
These `console.log`s should be removed as well.

FrankGalligan replied:
Done.

FrankGalligan commented:
OK, the history looks good.

FrankGalligan commented on README.md:
```diff
@@ -105,6 +108,13 @@ processGltf(gltf, options)
+|`--draco.quantizeTexcoord`|Quantization bits for texture coordinate attribute when using Draco compression.|No, default `12`|
+|`--draco.quantizeColor`|Quantization bits for color attribute when using Draco compression.|No, default `8`|
+|`--draco.quantizeSkin`|Quantization bits for skinning attribute (joint indices and joint weights) when using Draco compression.|No, default `12`|
+|`--draco.unifiedQuantization`|Quantize positions of all primitives using the same quantization grid defined by the unified bounding box of all primitives. If this option is not set, quantization is applied on each primitive separately which can result in gaps appearing between different primitives.|No, default `false`|
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+var Cesium = require('cesium');
+var hashObject = require('object-hash');
+var ForEach = require('./ForEach');
+var numberOfComponentsForType = require('./numberOfComponentsForType');
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+
+module.exports = compressMeshes;
+
+
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+    var componentReader = getComponentReader(accessor.componentType);
+
+    var values = [];
+    for (var i = 0; i < count; i++) {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+        componentReader(dataView, byteOffset, numberOfComponents, componentTypeByteLength, components);
+
+        for (var j = 0; j < numberOfComponents; ++j) {
+          values.push(components[j]);
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+
+
+function TypeToName() {
+}
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+
+
+
+
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,524 @@
+
+    return {
+      numComponents : componentsPerAttribute,
+      numVertices : accessor.count,
```
Done.

shehzan10 commented:
CI is failing on ESLint: https://travis-ci.org/AnalyticalGraphicsInc/gltf-pipeline/jobs/354149293#L1718-L1732

FrankGalligan commented:
@shehzan10 I fixed the lint errors in compressMeshes.js, but the error in updateVersion.js is in code I did not touch and a little harder to fix as it is the first parameter that is not used.

So I can add some code like
`if (!defined(accessorId)) {
    return;
}
`

Or I could add a new function to ForEach.js like
`ForEach.meshPrimitiveAttributeSemantic(primitive, function(semantic) {`

Any suggestions?

lilleyse commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+            var attributeToId = {};
+            var attributes = primitive.attributes;
+            for (var semantic in attributes) {
+                if (attributes.hasOwnProperty(semantic)) {
```
Use `ForEach.meshPrimitiveAttribute`

FrankGalligan replied:
Done.

lilleyse commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+            }
+
+            var encodedDracoDataArray = new encoderModule.DracoInt8Array();
+            encoder.SetSpeedOptions(10 - compressionLevel, 10 - compressionLevel);  // Compression level is 10 - speed.
```
Somewhere we should validate that the input options are in correct ranges. Cesium has utility functions for this like: `Check.typeOf.number.greaterThan('width', width, 0);` It will throw an error if the input is out of range.

FrankGalligan replied:
Added some checks to compressDracoMeshes.

lilleyse commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+
+module.exports = compressMeshes;
+
+function readAccessorPacked(gltf, accessor) {
```
Maybe move this to its own file.

FrankGalligan replied:
Done.

lilleyse commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    });
+
+    if (defined(gltf.extensionsUsed) &&
+        gltf.extensionsUsed.indexOf('KHR_draco_mesh_compression') >= 0) {
```
Use `hasExtension`.

FrankGalligan replied:
Done.

lilleyse commented:
For the eslint error in `updateVersion` maybe just put `/*eslint-disable no-unused-vars*/` on the line above:

```
ForEach.meshPrimitive(mesh, function(primitive) {
    /*eslint-disable no-unused-vars*/
    ForEach.meshPrimitiveAttribute(primitive, function(accessorId, semantic) {
```

lilleyse commented on README.md:
```diff
@@ -36,6 +36,9 @@ npm install -g gltf-pipeline
 `gltf-pipeline -i model.glb -j`
 
+#### Converting a glTF to Draco glTF
+`gltf-pipeline.js -i model.gltf -d -s -o modelDraco.gltf`
```
Ah never mind about this one.

FrankGalligan replied:
OK

TimvanScherpenzeel replied:
Will it stay required to run it with `-s`? I would like to embed the textures into the `.glb` file but it currently does not correctly do so when using it without `-s`.

TimvanScherpenzeel replied:
I was able to embed the textures correctly after using a `.gltf` instead of a `.glb` as input source. Using a `.glb` as input resulted in a `66kb` file instead of a `~3.3mb` file.

lilleyse replied:
Ok I see the problem now when using a glb input with `-d`. Perhaps a buffer is getting optimized away?

Good catch @TimvanScherpenzeel

TimvanScherpenzeel replied:
It appears that only the first buffer is being used, the other ones are dropped.

Broken:
<img width="317" alt="screen shot 2018-04-23 at 19 52 48" src="https://user-images.githubusercontent.com/11870513/39144181-5217ded8-4730-11e8-83d9-55de9faea87c.png">

Working:
<img width="334" alt="screen shot 2018-04-23 at 19 56 07" src="https://user-images.githubusercontent.com/11870513/39144220-6a80491a-4730-11e8-9cb9-c9bb30c22c6e.png">

FrankGalligan replied:
I'm having trouble reproducing the issue. With just a 2.0 branch I cannot open the DamagedHelmet.glb and output as a non-draco gltf file (or any file).

This is my command line:
node bin/gltf-pipeline.js -i ~/github/glTF-Sample-Models/2.0/DamagedHelmet/glTF-Binary/DamagedHelmet.glb 

TypeError: header.equals is not a function
    at getImageExtension (~/github/AnalyticalGraphicsInc/gltf-pipeline/lib/getImageExtension.js:18:16)
    at writeImage (~/github/AnalyticalGraphicsInc/gltf-pipeline/lib/writeResources.js:78:21)
    at ~/github/AnalyticalGraphicsInc/gltf-pipeline/lib/writeResources.js:48:23

TimvanScherpenzeel replied:
As an input example that reproduces the issue I've attached a .zip file containing the various outputs. The original `.gltf` input file (including seperate textures and `.bin`) can be found in `seperate-gltf`: [models.zip](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/files/1944328/models.zip)

`(embed-draco-glb)`
`node ./bin/gltf-pipeline.js -i /Users/TVS/Projects/vendor/gltf-pipeline/models/seperate-gltf/DamagedHelmet.gltf -d -b` results in a correctly outputted Draco `.glb` file (3.3 MB)

`(embed-glb)`
`node ./bin/gltf-pipeline.js -i /Users/TVS/Projects/vendor/gltf-pipeline/models/seperate-gltf/DamagedHelmet.gltf -b` results in a correctly outputted regular `.glb` file (3.8 MB)

`(embed-draco-broken-glb)`
`node ./bin/gltf-pipeline.js -i /Users/TVS/Projects/vendor/gltf-pipeline/models/embed-glb/DamagedHelmet.glb -d -b` results in a broken outputted `.glb` file (66 KB). Note that I'm using the `.glb` file as input.

FrankGalligan replied:
@TimvanScherpenzeel I fixed the issue. The bufferViews for inlined images were not being processed.

TimvanScherpenzeel replied:
Thanks! 👍

pjcozzi commented:
@FrankGalligan can you guys also send in a CLA?  This project uses the same CLA as Cesium: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla

shehzan10 commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+function compressMeshes(gltf, options) {
+    addExtensionsRequired(gltf, 'KHR_draco_mesh_compression');
+    var hashPrimitives = [];
+    options = options === undefined ? {} : options;
```
For better usage, I think this should be:
```
options = defaultValue(options, {});
options.dracoOptions = defaultValue(options.dracoOptions, {});
```

Otherwise the `options.dracoOptions.property` will result in `dracoOptions` being undefined.

FrankGalligan replied:
Done.

shehzan10 commented:
@FrankGalligan I've been running our tools using this draco PR. Other than the issues mentioned in the PR comments, I've noticed these models failing during draco encoding.

[draco-test.gltf.txt](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/files/1839151/draco-test.gltf.txt)
[draco-test-no-batchid.gltf.txt](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/files/1839152/draco-test-no-batchid.gltf.txt)

I've run both of these models through gltf-validator, and both pass.

Is this something you will be able to track down and debug?

CLI: `node ./bin/gltf-pipeline.js -i draco-test.gltf --compressMeshes`

Error:
```
Assertion failed: (normal.AbsSum()) <= (upper_bound), at: /google/src/cloud/ostava/fix/google3/third_party/draco/src/draco/compression/attributes/prediction_schemes/mesh_prediction_scheme_geometric_normal_predictor_area.h,95,ComputePredictedValue at Error
    at jsStackTrace (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:6:15080)
    at stackTrace (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:6:15251)
    at ___assert_fail (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:6:54271)
    at Ac (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:12:160232)
    at Array.nc (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:12:117615)
    at Array.jc (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:12:104961)
    at Array.Km (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:177359)
    at Array.Rj (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:147885)
    at Array._j (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:150645)
    at Array.vk (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:156918)
    at Array.Ve (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:10:187213)
    at zd (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:10:84649)
    at Yh (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:105774)
    at Oi (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:128196)
    at Kg (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:66952)
    at Vh (C:\workspace\gltf-pipeline-2.0\node_modules\draco3d\draco_encoder_nodejs.js:11:104198)
```

FrankGalligan commented on README.md:
```diff
@@ -105,6 +108,14 @@ processGltf(gltf, options)
 |`--separateTextures`, `-t`|Write out separate textures only.|No, default `false`|
 |`--checkTransparency`|Do a more exhaustive check for texture transparency by looking at the alpha channel of each pixel. By default textures are considered to be opaque.|No, default `false`|
 |`--stats`|Print statistics to console for input and output glTF files.|No, default `false`|
+|`--compressMesh`, `-d`|Compress all of the Meshes using Draco.|No, default `false`|
```
How about --compressDracoMesh?

Eventually I want to add support for animation.

FrankGalligan replied:
Ahh. OK I see below.

FrankGalligan commented on README.md:
```diff
@@ -105,6 +108,14 @@ processGltf(gltf, options)
 |`--separateTextures`, `-t`|Write out separate textures only.|No, default `false`|
 |`--checkTransparency`|Do a more exhaustive check for texture transparency by looking at the alpha channel of each pixel. By default textures are considered to be opaque.|No, default `false`|
 |`--stats`|Print statistics to console for input and output glTF files.|No, default `false`|
+|`--compressMesh`, `-d`|Compress all of the Meshes using Draco.|No, default `false`|
```
Done.

FrankGalligan commented on bin/gltf-pipeline.js:
```diff
@@ -181,7 +218,9 @@ var options = {
-    textureCompressionOptions : textureCompressionOptions
+    textureCompressionOptions : textureCompressionOptions,
+    compressMeshes: argv.compressMeshes,
+    dracoOptions: dracoOptions,
```
Done.

FrankGalligan commented on lib/updateVersion.js:
```diff
@@ -572,6 +572,7 @@ function stripAsset(gltf) {
 var knownExtensions = {
     CESIUM_RTC : true,
     KHR_materials_common : true,
+    KHR_draco_mesh_compression : true,
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+'use strict';
```
Renamed to compressDracoMeshes.js

FrankGalligan commented on lib/processGltf.js:
```diff
@@ -90,6 +91,11 @@ function getStages(options) {
+
+    // TODO: Add quantization.
+    if (options.compressMeshes) {
+      stages.push(compressMeshes);
```
Done.

FrankGalligan commented on bin/gltf-pipeline.js:
```diff
@@ -181,7 +218,9 @@ var options = {
-    textureCompressionOptions : textureCompressionOptions
+    textureCompressionOptions : textureCompressionOptions,
+    compressMeshes: argv.compressMeshes,
+    dracoOptions: dracoOptions,
```
Done.

FrankGalligan commented on lib/processGltf.js:
```diff
@@ -10,6 +10,7 @@ var removeDefaults = require('./removeDefaults');
 var removePipelineExtras = require('./removePipelineExtras');
 var updateVersion = require('./updateVersion');
 var writeResources = require('./writeResources');
+var compressMeshes = require('./compressMeshes');
```
Done.

FrankGalligan commented on bin/gltf-pipeline.js:
```diff
@@ -114,7 +114,40 @@ var argv = yargs
+        },
+        'compressMeshes': {
+            alias: 'd',
+            describe: 'Compress the geometries using mesh compression extension.',
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+var getAccessorByteStride = require('./getAccessorByteStride');
+var getComponentReader = require('./getComponentReader.js');
+var hashObject = require('object-hash');
+var numberOfComponentsForType = require('./numberOfComponentsForType');
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+ */
+function Remove() {}
+
+Remove.accessor = function(gltf, accessorId) {
```
created file removedUnusedElements.js. For now this file exports one function that will remove "all" unused elements. This could always be refactored later if other code needed to call remove on one type at a time.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+
+    ForEach.mesh(gltf, function(mesh) {
+        ForEach.meshPrimitive(mesh, function(primitive) {
+            ForEach.meshPrimitiveAttribute(primitive, function(attributeAccessorId, semantic) {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    });
+
+    if (defined(gltf.extensionsUsed) &&
+        gltf.extensionsUsed.indexOf('Draco_animation_compression') >= 0) {
```
No, I removed all of the code reference Draco animation compression. That is something that we will add later.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+
+    if (defined(gltf.extensionsUsed) &&
+        gltf.extensionsUsed.indexOf('Draco_animation_compression') >= 0) {
+        gltf.extensions.Draco_animation_compression.forEach(function (compressedAnimation) {
```
Maybe, but there is much less code now that I removed the animation code. I can revisit when adding the animation code later.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+        gltf.extensions.Draco_animation_compression.forEach(function (compressedAnimation) {
+            compressedAnimation.bufferView--;
+        });
+    }
```
Removed.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    }
+    var draco_extension = {
+        bufferView : bufferViewId,
+        attributes : attributeToId,
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    // Calculate accessor's that are currently in use.
+    var usedAccessorIds = {};
+
+    ForEach.mesh(gltf, function(mesh) {
```
I added the support for removing an acccessor for files with morph targets, but only for the mesh compression.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    return 'bufferViews';
+};
+
+function removeUnusedElements(gltf, type, usedIds) {
```
I created a removeUnusedElements.js file and refactored the code a lot. Take a look and let me know if you want me to change anything.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+}
+
+function addCompressionExtensionToPrimitive(gltf, primitive,
+    attributeToId, encodedLen, encodedData) {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+}
+
+function addCompressionExtensionToPrimitive(gltf, primitive,
+    attributeToId, encodedLen, encodedData) {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+        extras : {
+            _pipeline : {
+                extension : '.bin',
+                source :encodedData
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+        byteLength : encodedLen,
+        extras : {
+            _pipeline : {
+                extension : '.bin',
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    primitive.extensions = extensions;
+    var copied_extension = {
+        bufferView : draco_extension.bufferView,
+        attributes : draco_extension.attributes,
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+
+function copyCompressedExtensionToPrimitive(primitive, compressedPrimitive) {
+    var attributes = {};
+    for (var semantic in primitive.attributes) {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+                mode : primitive.mode
+            };
+            var hashValue = hashObject(primitiveGeometry);
+            if (hashPrimitives[hashValue] !== undefined) {
```
Done.

lilleyse commented on lib/compressDracoMeshes.js:
```diff
@@ -0,0 +1,286 @@
+    addExtensionsRequired(gltf, 'KHR_draco_mesh_compression');
+    var hashPrimitives = [];
+    options = defaultValue(options, {});
+    options.dracoOptions= defaultValue(options.dracoOptions, {});
```
Small tweak: `options.dracoOptions =`

FrankGalligan replied:
Done.

lilleyse commented on lib/compressDracoMeshes.js:
```diff
@@ -0,0 +1,286 @@
+
+            // Add attributes to mesh.
+            var attributeToId = {};
+            var attributes = primitive.attributes;
```
This variable is unused.

FrankGalligan replied:
Done.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+
+/**
+ * Removes unused elements from gltf.
+ * This function currenlty only works for accessors, buffers, and bufferViews.
```
`currenlty` typo

FrankGalligan replied:
Done.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+ * Contains functions for getting a list of element ids in use by the glTF asset.
+ * @constructor
+ */
+function getListOfElementsIdsInUse() {}
```
Looks slightly clearer if `getListOfElementsIdsInUse` is capitalized.

FrankGalligan replied:
So do you want it as 'GetListOfElementsIdsInUse', or 'getListOfElementsIDsInUse', or something else?

lilleyse replied:
Actually it's probably fine as is.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+/**
+ * Contains functions for getting a list of element ids in use by the glTF asset.
+ * @constructor
+ */
```
For each of the new classes mark as `@private` so they don't show up in the doc.

FrankGalligan replied:
Done.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+
+TypeToGltfElementName.bufferView = function() {
+    return 'bufferViews';
+};
```
`TypeToGltfElementName` can be simplified to

```
var TypeToGltfElementName = {
    accessor : 'accessors',
    buffer : 'buffers',
    bufferView : 'bufferViews'
};
```

FrankGalligan replied:
Done.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+ * This function currenlty only works for accessors, buffers, and bufferViews.
+ *
+ * @param {Object} gltf A javascript object containing a glTF asset.
+ * @returns {Object} The glTF asset with the added pipeline extras.
```
Fix the description, looks like a copy-paste error.

FrankGalligan replied:
Done.

pjcozzi commented:
We received a CLA from @FrankGalligan, thanks again!

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+    ForEach.skin(gltf, function(skin) {
+       if (defined(skin.inverseBindMatrices)) {
+           usedAccessorIds[skin.inverseBindMatrices] = true;
+       }
```
The indentation in this block is slightly off.

FrankGalligan replied:
Done.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -0,0 +1,235 @@
+'use strict';
```
Does the uncompressed data get spliced out of the original buffer in all cases?

It's possible that the uncompressed attributes and a texture may be stored in a buffer, and because that texture's bufferView references the buffer the buffer won't get removed in `removeUnusedElements`.

FrankGalligan replied:
Currently yes. I have another branch that I think should fix this issue here:
https://github.com/FrankGalligan/gltf-pipeline/commits/fix_buffer_packing

I can merge that branch into this branch. I also have some other branches as well that I want to merge. Like a fix for JOINT_ATTRIBUTES not using integer, and a branch that will add support to go from Draco glTF -> glTF.

I was going to make PRs from them after this was landed as I think it would be easier to review, but I can merge them into this this PR if you want.

lilleyse replied:
Maybe open a PR for `fix_buffer_packing` that targets this branch? The others can wait for later.

FrankGalligan replied:
https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/363

FrankGalligan commented:
@shehzan10 I fixed the issue you were having with the encoder. I updated the npm package, so if you update your dependencies this should compress now.  There isstill an issue with the writing of the gltf, which I'm looking into.

shehzan10 commented:
Can confirm that the updates fix the crashing models.

> There is still an issue with the writing of the gltf, which I'm looking into.

Writing to disk looks ok, I guess. But, it looks like the test model fails to load for compression level 7 and above and works for 6 and lower.

FrankGalligan commented:
@shehzan10 Yeah so the decoder has the dcheck as well. It the decoder is using the nodejs modules and they update it should be fine. Otherwise they will have to wait until we push the fixes to github. We are not pushing the changes until the week of 4/16.

shehzan10 commented:
@FrankGalligan can you update this to draco 1.3 and fix the remaining comments from @lilleyse? I think we should be able to merge then.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+        extensions = {};
+        primitive.extensions = extensions;
+    }
+    var draco_extension = {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    var draco_extension = compressedPrimitive.extensions.KHR_draco_mesh_compression;
+    var extensions = {};
+    primitive.extensions = extensions;
+    var copied_extension = {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+    addExtensionsRequired(gltf, 'KHR_draco_mesh_compression');
+    var hashPrimitives = [];
+    options = options === undefined ? {} : options;
+    var compressionLevel = !defined(options.dracoOptions.compressionLevel) ? 7 : options.dracoOptions.compressionLevel;
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+        var accessors = gltf.accessors;
+        var min = arrayFill(new Array(3), Number.POSITIVE_INFINITY);
+        var max = arrayFill(new Array(3), Number.NEGATIVE_INFINITY);
+        ForEach.accessorWithSemantic(gltf, "POSITION", function(accessorId) {
```
Done.

FrankGalligan commented on lib/compressMeshes.js:
```diff
@@ -0,0 +1,518 @@
+                return;
+            }
+
+            // TODO: Use hash map.
```
Done.

FrankGalligan commented on bin/gltf-pipeline.js:
```diff
@@ -114,7 +114,41 @@ var argv = yargs
+            default : false,
+            describe: 'Quantize positions of all primitives using the same quantization grid defined by the unified bounding box of all primitives. If this option is not set, quantization is applied on each primitive separately which can result in gaps appearing between different primitives. Default is false.',
+            type: 'boolean'
+        },
```
Done.

FrankGalligan commented:
@shehzan10 Updated to 1.3. Waiting on one comment form @lilleyse

lilleyse commented:
The code changes looks good. We're just going to do some final testing.

lilleyse commented:
All good here - thanks @FrankGalligan
